include/marl/deprecated.h

```diff
@@ -0,0 +1,36 @@
+// Copyright 2020 The Marl Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef marl_deprecated_h
+#define marl_deprecated_h
+
+// Deprecated marl::Scheduler methods:
+//   Scheduler(Allocator* allocator = Allocator::Default)
+//   getThreadInitializer(), setThreadInitializer()
+//   getWorkerThreadCount(), setWorkerThreadCount()
+#ifndef MARL_ENABLE_DEPRECATED_SCHEDULER_GETTERS_SETTERS
+#define MARL_ENABLE_DEPRECATED_SCHEDULER_GETTERS_SETTERS 1
+#endif  // MARL_ENABLE_DEPRECATED_SCHEDULER_GETTERS_SETTERS
+
+#ifndef MARL_WARN_DEPRECATED
+#define MARL_WARN_DEPRECATED 0
+#endif  // MARL_WARN_DEPRECATED
+
+#if MARL_WARN_DEPRECATED
+#define MARL_DEPRECATED(message) __attribute__((deprecated(message)))
```

Comment:
So one thing about deprecation that's important is putting a deadline on when deprecated functions are going to be removed, and making sure to let the user know. For example, some projects remove deprecated functions at the next minor revision change (e.g. UE4 does that). In that case, you could make this something like:

```
#define MARL_DEPRECATED(message) __attribute__((deprecated(message " This will be removed at the next minor revision upgrade.")))
```

Or you can actually specify the version in the message. If you don't do so, then clients will either ignore the deprecation warning, or be annoyed when upgrading the library.

Reply:
Yes. I have a plan to gradually roll this out:

1. Land this change that adds the new API, but does not warn about usage of the old API. 
2. Have this propagate into all of Google's projects and migrate Google usages to the new API. This will confirm that the new API is good, and gives me an opportunity to make changes before encouraging usage of the new API.
3. Land [this branch](https://github.com/ben-clayton/marl/tree/sched-cfg-2) to switch all tests, examples, benchmarks, documentation over to the new API.
4. File a separate issue to announce the deprecation, along with planned soft & hard dates of removal.
5. Land [this branch](https://github.com/ben-clayton/marl/tree/sched-cfg-3) that enables the warning that the old API is deprecated.
6. Wait for the soft deprecation deadline to be reached and land the change that disables the old API by default.
7. Wait for the hard deprecation deadline to be reached, and strip out the old API from the codebase.

Reply:
Note at 3. I'll also include a change to link to the github issue from the `MARL_DEPRECATED` macro.